src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Android.cs

```diff
@@ -0,0 +1,390 @@
+            {
+                e = new InvalidTimeZoneException(SR.Format(SR.InvalidTimeZone_InvalidFileData, id, GetTimeZoneDirectory() + _timeZoneFileName));
+                return TimeZoneInfoResult.TimeZoneNotFoundException;
+            }
+
+            return TimeZoneInfoResult.Success;
+        }
+
+        private static string[] GetTimeZoneIds()
+        {
+            return s_androidTzDataInstance.GetTimeZoneIds();
+        }
+
+        /*
+        * Android v4.3 Timezone support infrastructure.
+        *
+        * This is a C# port of libcore.util.ZoneInfoDB:
+        *
+        *    https://android.googlesource.com/platform/libcore/+/master/luni/src/main/java/libcore/util/ZoneInfoDB.java
+        *
+        * This is needed in order to read Android v4.3 tzdata files.
+        *
+        * Android 10+ moved the up-to-date tzdata location to a module updatable via the Google Play Store and the
+        * database location changed (https://source.android.com/devices/architecture/modular-system/runtime#time-zone-data-interactions)
+        * The older locations still exist (at least the `/system/usr/share/zoneinfo` one) but they won't be updated.
+        */
+        private sealed class AndroidTzData
+        {
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataHeader
+            {
+                public fixed byte signature[12]; // "tzdata2012f\0"
+                public int indexOffset;
+                public int dataOffset;
+                public int finalOffset;
+            }
+
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataEntry
+            {
+                public fixed byte id[40];
+                public int byteOffset;
+                public int length;
+                public int unused; // Was raw GMT offset; always 0 since tzdata2014f (L).
+            }
+
+            private string[]? _ids;
+            private int[]? _byteOffsets;
+            private int[]? _lengths;
+
+            public AndroidTzData()
+            {
+                ReadHeader(GetTimeZoneDirectory() + _timeZoneFileName);
+            }
+
+            private unsafe void ReadHeader(string tzFilePath)
+            {
+                int size = Math.Max(Marshal.SizeOf(typeof(AndroidTzDataHeader)), Marshal.SizeOf(typeof(AndroidTzDataEntry)));
+                var buffer = new byte[size];
+                AndroidTzDataHeader header = ReadAt<AndroidTzDataHeader>(tzFilePath, 0, buffer);
+
+                header.indexOffset = NetworkToHostOrder(header.indexOffset);
+                header.dataOffset = NetworkToHostOrder(header.dataOffset);
+
+                // tzdata files are expected to start with the form of "tzdata2012f\0" depending on the year of the tzdata used which is 2012 in this example
+                sbyte* s = (sbyte*)header.signature;
+                string magic = new string(s, 0, 6, Encoding.ASCII);
```

Comment:
Do we really need to create a string just to compare the `magic` header?

Reply:
What's the preferred way to see if `byte signature[12]` begins with "tzdata" in bytes?
Would it be better to create the byte form of "tzdata" via `Encoding.ASCII.GetBytes("tzdata")` as a variable and iterate through to compare?

Reply:
Those approaches all involve allocating for no reason.

One easy approach is you can just compare each individual byte:

```C#
if (header.signature[0] != (byte)'t' || header.signature[0] != (byte)'z' || ...
```

Or you can just hard code a larger int, and compare multiple bytes at a time:

https://github.com/dotnet/runtime/blob/164a0c079a44ec82a2af71767b1eefb255242aa9/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs#L1758-L1764

Note - here you need 6 bytes, so you'd probably use an `int` and a `short`.

Reply:
Thank you for the suggestions!

Reply:
After some changes with @steveisok 's help, we are allocating 8 bytes to utilize `TZif_ToInt64`, ignoring the last two bytes when comparing to `tzdata`.